elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/ElasticsearchSourceStage.scala

```diff
@@ -0,0 +1,124 @@
+
+final case class OutgoingMessage[T](id: String, source: T)
+
+final class ElasticsearchSourceStage[T](indexName: String,
+                                        typeName: String,
+                                        query: String,
+                                        client: RestClient,
+                                        settings: ElasticsearchSourceSettings)(implicit reader: JsonReader[T])
+    extends GraphStage[SourceShape[OutgoingMessage[T]]] {
+
+  val out: Outlet[OutgoingMessage[T]] = Outlet("ElasticsearchSource.out")
+  override val shape: SourceShape[OutgoingMessage[T]] = SourceShape(out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new ElasticsearchSourceLogic[T](indexName, typeName, query, client, settings, out, shape)
+
+}
+
+sealed class ElasticsearchSourceLogic[T](indexName: String,
+                                         typeName: String,
+                                         query: String,
+                                         client: RestClient,
+                                         settings: ElasticsearchSourceSettings,
+                                         out: Outlet[OutgoingMessage[T]],
+                                         shape: SourceShape[OutgoingMessage[T]])(implicit reader: JsonReader[T])
+    extends GraphStageLogic(shape)
+    with ResponseListener {
+
+  private var started = false
+  private var scrollId: String = null
+  private val responseHandler = getAsyncCallback[Response](handleResponse)
+  private val failureHandler = getAsyncCallback[Throwable](handleFailure)
+
+  def sendScrollScanRequest(): Unit =
+    try {
+      if (scrollId == null) {
+        client.performRequestAsync(
+          "POST",
+          s"$indexName/$typeName/_search",
+          Map("scroll" -> "5m", "sort" -> "_doc").asJava,
+          new StringEntity(s"""{"size": ${settings.bufferSize}, "query": ${query}}"""),
+          this
+        )
+      } else {
+        client.performRequestAsync(
+          "POST",
+          s"/_search/scroll",
+          Map[String, String]().asJava,
+          new StringEntity(Map("scroll" -> "5m", "scroll_id" -> scrollId).toJson.toString),
+          this
+        )
+      }
+    } catch {
+      case ex: Exception => handleFailure(ex)
+    }
+
+  override def onFailure(exception: Exception) = failureHandler.invoke(exception)
+  override def onSuccess(response: Response) = responseHandler.invoke(response)
+
+  def handleFailure(ex: Throwable): Unit =
+    failStage(ex)
+
+  def handleResponse(res: Response): Unit = {
+    val json = {
+      val out = new ByteArrayOutputStream()
+      try {
+        res.getEntity.writeTo(out)
+        new String(out.toByteArray, "UTF-8")
+      } finally {
+        out.close()
+      }
+    }
+
+    val jsObj = json.parseJson.asJsObject
+
+    jsObj.fields.get("error") match {
+      case None => {
+        val hits = jsObj.fields("hits").asJsObject.fields("hits").asInstanceOf[JsArray]
```

Comment:
It would be possible. But the REST client of Elasticsearch is improving rapidly. I expect it will become to return Java models instead of json string as same as the transport client (it's an another type of Elasticsearch client) in the future. So now creating our models might be overinvestment.